src/backend/opengl/CommandBufferGL.cpp

```diff
@@ -446,85 +374,249 @@ namespace backend { namespace opengl {
+                // this function) may be undefined for other texture formats.
+                ASSERT(format == nxt::TextureFormat::D32FloatS8Uint);
+            }
+        }
+
+        // Clear framebuffer attachments as needed
+        {
+            for (uint32_t i : IterateBitSet(renderPass->GetColorAttachmentMask())) {
+                const auto& attachmentInfo = renderPass->GetColorAttachment(i);
+
+                // Load op - color
+                if (attachmentInfo.loadOp == nxt::LoadOp::Clear) {
+                    glClearBufferfv(GL_COLOR, i, attachmentInfo.clearColor.data());
+                }
+            }
+
+            if (renderPass->HasDepthStencilAttachment()) {
+                const auto& attachmentInfo = renderPass->GetDepthStencilAttachment();
+                nxt::TextureFormat attachmentFormat =
+                    attachmentInfo.view->GetTexture()->GetFormat();
+
+                // Load op - depth/stencil
+                bool doDepthClear = TextureFormatHasDepth(attachmentFormat) &&
+                                    (attachmentInfo.depthLoadOp == nxt::LoadOp::Clear);
+                bool doStencilClear = TextureFormatHasStencil(attachmentFormat) &&
+                                      (attachmentInfo.stencilLoadOp == nxt::LoadOp::Clear);
+                if (doDepthClear && doStencilClear) {
+                    glClearBufferfi(GL_DEPTH_STENCIL, 0, attachmentInfo.clearDepth,
+                                    attachmentInfo.clearStencil);
+                } else if (doDepthClear) {
+                    glClearBufferfv(GL_DEPTH, 0, &attachmentInfo.clearDepth);
+                } else if (doStencilClear) {
+                    const GLint clearStencil = attachmentInfo.clearStencil;
+                    glClearBufferiv(GL_STENCIL, 0, &clearStencil);
+                }
+            }
+        }
+
+        RenderPipeline* lastPipeline = nullptr;
+        uint32_t indexBufferBaseOffset = 0;
+
+        PersistentPipelineState persistentPipelineState;
+
+        PushConstantTracker pushConstants;
+        InputBufferTracker inputBuffers;
+
+        // Set defaults for dynamic state
+        persistentPipelineState.SetDefaultState();
+        glBlendColor(0, 0, 0, 0);
+        glViewport(0, 0, renderPass->GetWidth(), renderPass->GetHeight());
+        glScissor(0, 0, renderPass->GetWidth(), renderPass->GetHeight());
+
+        Command type;
+        while (mCommands.NextCommandId(&type)) {
+            switch (type) {
+                case Command::EndRenderPass: {
+                    mCommands.NextCommand<EndRenderPassCmd>();
+                    glDeleteFramebuffers(1, &fbo);
+                    return;
+                } break;
+
                 case Command::DrawArrays: {
                     DrawArraysCmd* draw = mCommands.NextCommand<DrawArraysCmd>();
-                    pushConstants.Apply(lastPipeline, lastGLPipeline);
+                    pushConstants.Apply(lastPipeline, lastPipeline);
```

Comment:
Apply() always seems to be called with lastPipeline twice now (although I could be wrong; didn't check non-diffed code). Consider removing one of the params if possible?

Reply:
Never mind.. I see now this is necessary to force two different upcasts. I hate MI...

Reply:
Whoops sorry I forgot to answer to this. Yeah that's because of multiple inheritance :/.